File: c++/datastructures/BinaryTree.cpp

#include "BinaryTree.hpp"
#include <queue>

using namespace std;

Node::Node(int data)
{
    this->data = data;
    this->left = this->right = nullptr;
}

Node::Node(int data, Node* left, Node* right)
{
    this->data = data;
    this->left = left;
    this->right = right;
}

static int idx = -1;

Node* buildTree(const vector<int>& preorder)
{
    idx++;
    if(preorder[idx] == -1) return nullptr;
    Node* root = new Node(preorder[idx]);
    root->left = buildTree(preorder);
    root->right = buildTree(preorder);
    return root;
}

void preorder_traversal(Node* root)
{
    if(root == nullptr) return;
    cout << root->data << " ";
    preorder_traversal(root->left);
    preorder_traversal(root->right);
}

void postorder_traversal(Node* root)
{
    if(root == nullptr) return;
    postorder_traversal(root->left);
    postorder_traversal(root->right);
    cout << root->data << " ";
}

void inorder_traversal(Node* root)
{
    if(root == nullptr) return;
    inorder_traversal(root->left);
    cout << root->data << " ";
    inorder_traversal(root->right);
}

void levelorder_traversal(Node* root)
{
    queue<Node*> q;
    q.push(root);
    while(q.size() > 0)
    {
        Node* current = q.front();
        q.pop();
        cout << current->data << " ";
        if (current->left != nullptr)
        {
            q.push(current->left);
        }
        if (current->right != nullptr)
        {
            q.push(current->right);
        }
    }
}

void print_tree(Node *root, Mode mode)
{
    switch (mode)
    {
        case PREORDER: preorder_traversal(root); cout << "\n"; break;
        case POSTORDER: postorder_traversal(root); cout << "\n"; break;
        case INORDER: inorder_traversal(root); cout << "\n"; break;
        case LEVELORDER: levelorder_traversal(root); cout << "\n"; break;
    }
}

File: python/datastructures/BinaryTree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

idx = -1
def insert(seq):
    global idx
    idx += 1
    if idx >= len(seq) or seq[idx] is None:
        return None
    root = Node(seq[idx])
    root.left = insert(seq)
    root.right = insert(seq)
    return root

def preorder_traversal(root):
    if root is None:
        return
    print(root.data, end=" ")
    preorder_traversal(root.left)
    preorder_traversal(root.right)

if __name__ == "__main__":
    seq = [1, 2, None, None, 3, 4, None, None, 5, None, None]
    root = insert(seq)
    preorder_traversal(root)